### PR TITLE
*: support granting a role to a user in v3 auth

### PR DESCRIPTION
**Pull request by mitake**

### CODE REVIEW

File: auth/store.go [hunk 1]
```diff
@@ -34,10 +35,11 @@ var (
 
 	plog = capnslog.NewPackageLogger("github.com/coreos/etcd", "auth")
 
-	ErrUserAlreadyExist = errors.New("auth: user already exists")
-	ErrUserNotFound     = errors.New("auth: user not found")
-	ErrRoleAlreadyExist = errors.New("auth: role already exists")
-	ErrRoleNotFound     = errors.New("auth: role not found")
+	ErrUserAlreadyExist   = errors.New("auth: user already exists")
+	ErrUserNotFound       = errors.New("auth: user not found")
+	ErrUserAlreadyGranted = errors.New("auth: a user is already granted a role")
```
Comment (xiang90) [inline auth/store.go]: ErrRoleAlreadyGranted? auth: role is already granted to the user. (do we really need to make this an error? double grant seems to be just fine?)

File: auth/store.go [hunk 2]
```diff
@@ -56,6 +58,9 @@ type AuthStore interface {
 	// UserChangePassword changes a password of a user
 	UserChangePassword(r *pb.AuthUserChangePasswordRequest) (*pb.AuthUserChangePasswordResponse, error)
 
+	// UserGrant changes a password of a user
```
Comment (xiang90) [inline auth/store.go]: update the comment UserGrant grants a role to the user

File: auth/store.go [hunk 3]
```diff
@@ -174,6 +179,46 @@ func (as *authStore) UserChangePassword(r *pb.AuthUserChangePasswordRequest) (*p
...
+	sort.Sort(sort.StringSlice(user.Roles))
+
+	marshaledUser, merr := user.Marshal()
+	if merr != nil {
+		return nil, merr
+	}
+
+	tx.UnsafePut(authUsersBucketName, user.Name, marshaledUser)
+
+	plog.Noticef("granted a role %s to a user %s", r.Role, r.User)
```
Comment (xiang90) [inline auth/store.go]: granted role %x to user %s (remove `a`)

**Comment by xiang90** [comment]
LGTM after fixing the comments.

File: auth/store.go (hunk 1 above)
Comment (mitake) [inline auth/store.go]: OK, I'll change the checking logic for simply ignoring the request with info logging.

File: auth/store.go (hunk 2 above)
Comment (mitake) [inline auth/store.go]: Oops, sorry for that!

File: auth/store.go (hunk 3 above)
Comment (mitake) [inline auth/store.go]: OK

File: auth/store.go (hunk 1 above)
Comment (xiang90) [inline auth/store.go]: probably just a warning at server side?

**Comment by mitake** [comment]
@xiang90 updated for the comments, PTAL.

File: auth/store.go (hunk 1 above)
Comment (xiang90) [inline auth/store.go]: then shall we remove this error from here?

File: auth/store.go [hunk 4]
```diff
@@ -174,6 +178,47 @@ func (as *authStore) UserChangePassword(r *pb.AuthUserChangePasswordRequest) (*p
...
+	}
+
+	_, vs = tx.UnsafeRange(authRolesBucketName, []byte(r.Role), nil, 0)
+	if len(vs) != 1 {
+		return nil, ErrRoleNotFound
+	}
+
+	idx := sort.SearchStrings(user.Roles, r.Role)
+	if idx < len(user.Roles) && strings.Compare(user.Roles[idx], r.Role) == 0 {
+		plog.Infof("user %s is already granted role %s", r.User, r.Role)
```
Comment (xiang90) [inline auth/store.go]: warnf?

**Comment by xiang90** [comment]
LGTM. I will get this merged tomorrow morning in US time. Thanks!

**Comment by mitake** [comment]
@xiang90 updated for the last comment. Thanks a lot for your review, good night :)

**Pull request merged**